src/teacher.py

```diff
@@ -77,23 +77,30 @@ def _answer_Demonstration_query(
 
 
 @dataclass
-class TeacherInteractionMonitor(utils.Monitor):
-    """Wraps the interaction between agent and teacher to include generating
-    and answering queries."""
+class TeacherInteractionMonitor(utils.VideoMonitor):
+    """A monitor that renders each state and action encountered and queries and
+    responses during interaction with the teacher.
+
+    The render_fn is generally env.render.
+    """
     _request: InteractionRequest
     _teacher: Teacher
     _responses: List[Optional[Response]] = field(init=False,
                                                  default_factory=list)
     _query_cost: float = field(init=False, default=0.0)
 
     def observe(self, state: State, action: Optional[Action]) -> None:
-        del action  # unused
         query = self._request.query_policy(state)
         if query is None:
-            self._responses.append(None)
+            response = None
+            caption = "None"
         else:
-            self._responses.append(self._teacher.answer_query(state, query))
+            response = self._teacher.answer_query(state, query)
             self._query_cost += query.cost
+            caption = f"{response}, cost={query.cost}"
+        self._responses.append(response)
+        if CFG.make_interaction_videos:
+            self._video.extend(self._render_fn(action, caption))
```

Comment:
@tomsilver are we 1000% sure that we want to be using the stateful render function here? We are given a state in this function, and I don't see reset being called anywhere. So I think the other one is the right one?

Reply:
yeah we want the stateful renderer. for example, pybullet environments will render based on their current internal state (which can include things that are not expressed in the State).

Reply:
yup, verified the state will always equal the `env._current_state`